File: src/cli/update.js
#!/usr/bin/env node

/**
 * Update an installed project from the packaged template files.
 */

import { execSync } from "node:child_process";
import {
  copyFileSync,
  existsSync,
  readdirSync,
  readFileSync,
  statSync,
  mkdirSync,
  writeFileSync,
} from "node:fs";
import { createRequire } from "node:module";
import { dirname, join, resolve, relative } from "node:path";
import { createInterface } from "node:readline/promises";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const require = createRequire(import.meta.url);

const COMMON_MANIFEST = [
  { src: "lib.typ", dest: "lib.typ", autoUpdate: false },
  { src: "flake.nix", dest: "flake.nix", autoUpdate: false },
  { src: "bibliography.bib", dest: "bibliography.bib", autoUpdate: false },
  { src: "README.md", dest: "README.md", autoUpdate: false },
  { src: ".prepare.config", dest: ".prepare.config", autoUpdate: false },
  { src: "prepare.js", dest: "prepare.js", autoUpdate: true },
  { src: "capture.js", dest: "capture.js", autoUpdate: true },
];

const HELP_TEXT = `
Lab Report Template Updater

Usage:
  node scripts/update.js [options]
  lab-report-update [options]

Options:
  --dest <path>   Destination directory (default: current working directory)
  --multi         Force multi-lab update mode
  -f, --force     Apply all updates without prompting
  -h, --help      Show this help message

Examples:
  lab-report-update --dest ~/projects/lab-01
  node scripts/update.js --force --multi
`;

class Updater {
  constructor(sourceDir, destDir, force, multiLab) {
    this.sourceDir = sourceDir;
    this.destDir = destDir;
    this.autoAcceptAll = force;
    this.multiLab = multiLab;
    this.appliedCount = 0;
    this.rl = createInterface({ input: process.stdin, output: process.stdout });
  }

  getFilesRecursively(dir) {
    const files = [];
    const read = (currentDir) => {
      for (const entry of readdirSync(currentDir, { withFileTypes: true })) {
        const fullPath = join(currentDir, entry.name);
        if (entry.isDirectory()) read(fullPath);
        else files.push(fullPath);
      }
    };
    read(dir);
    return files;
  }

  buildManifest() {
    const configPath = join(this.destDir, "labreport.json");
    let isMulti = this.multiLab;

    if (existsSync(configPath)) {
      try {
        const config = JSON.parse(readFileSync(configPath, "utf-8"));
        isMulti = isMulti || config.multiLab;
      } catch (e) {
        console.warn(`Error reading labreport.json: ${e.message}`);
      }
    } else {
      const defaultConfig = { multiLab: isMulti };
      writeFileSync(configPath, JSON.stringify(defaultConfig, null, 2));
      console.warn(
        `Warning: labreport.json not found in ${this.destDir}. Created default config.`,
      );
      console.warn("Please verify the configuration and run the command again.");
      process.exit(1);
    }

    console.log(`Detected ${isMulti ? "multi-lab" : "single-lab"} setup.`);
    const manifest = [...COMMON_MANIFEST];

    if (isMulti) {
      const readmeIndex = manifest.findIndex((m) => m.dest === "README.md");
      if (readmeIndex !== -1) {
        manifest[readmeIndex] = {
          ...manifest[readmeIndex],
          src: "multi/README.md",
        };
      }

      const subDirs = readdirSync(this.destDir).filter((f) => {
        try {
          return (
            statSync(join(this.destDir, f)).isDirectory() &&
            !f.startsWith(".") &&
            f !== "node_modules"
          );
        } catch {
          return false;
        }
      });

      for (const dir of subDirs) {
        if (existsSync(join(this.destDir, dir, "report.typ"))) {
          manifest.push({
            src: "multi/report.typ",
            dest: `${dir}/report.typ`,
            autoUpdate: false,
          });
        }
      }

      const fixedImgDir = join(this.sourceDir, "img", "fixed");
      if (existsSync(fixedImgDir)) {
        const files = this.getFilesRecursively(fixedImgDir);
        for (const file of files) {
          const relPath = relative(fixedImgDir, file).replace(/\\/g, "/");
          manifest.push({
            src: `img/fixed/${relPath}`,
            dest: `img/fixed/${relPath}`,
            autoUpdate: false,
          });
        }
      }
    } else {
      manifest.push({
        src: "report.typ",
        dest: "report.typ",
        autoUpdate: false,
      });
    }

    return manifest;
  }

  async processUpdates() {
    const manifest = this.buildManifest();
    console.log(`Checking for updates in: ${this.destDir}\n`);

    try {
      for (const item of manifest) {
        const srcPath = join(this.sourceDir, item.src);
        const dstPath = join(this.destDir, item.dest);
        const displayLabel = item.dest;

        if (!existsSync(srcPath)) continue;

        const isNew = !existsSync(dstPath);
        const hasChanged =
          !isNew &&
          !readFileSync(srcPath).equals(readFileSync(dstPath));

        if (isNew || hasChanged) {
          await this.handleFileUpdate(
            item,
            srcPath,
            dstPath,
            displayLabel,
            isNew,
          );
        }
      }
    } finally {
      this.rl.close();
    }

    console.log(`\nUpdate finished. ${this.appliedCount} files updated.`);
  }

  async handleFileUpdate(item, srcPath, dstPath, label, isNew) {
    const applyUpdate = () => {
      const dstDir = dirname(dstPath);
      if (!existsSync(dstDir)) {
        mkdirSync(dstDir, { recursive: true });
      }
      copyFileSync(srcPath, dstPath);
      this.appliedCount++;
    };

    if (item.autoUpdate || this.autoAcceptAll) {
      applyUpdate();
      const tag = item.autoUpdate ? "[AUTO-UPDATE]" : "[FORCED]";
      console.log(`✓ Updated ${tag}: ${label}`);
      return;
    }

    if (isNew) {
      console.log(`\n[NEW FILE] ${label}`);
    } else {
      this.showSmartDiff(label, dstPath, srcPath);
    }

    let shouldUpdate = false;
    while (true) {
      const answer = await this.rl.question(
        `Apply update to ${label}? (y)es / (n)o / (a)ll / (q)uit: `,
      );
      const choice = answer.toLowerCase().trim();

      if (choice === "a") {
        this.autoAcceptAll = true;
        shouldUpdate = true;
        break;
      } else if (choice === "y") {
        shouldUpdate = true;
        break;
      } else if (choice === "n") {
        break;
      } else if (choice === "q") {
        console.log("\nStopping update process.");
        this.rl.close();
        process.exit(0);
      }
    }

    if (shouldUpdate) {
      applyUpdate();
      console.log(`✓ Updated: ${label}`);
    } else {
      console.log(`Skipped: ${label}`);
    }
  }

  showSmartDiff(fileName, oldPath, newPath) {
    if (fileName.match(/\.(png|jpe?g|gif|svg|webp|ico)$/i)) {
      console.log(`\n--- Diff for ${fileName} ---`);
      console.log("  [Binary image file changed. Cannot display inline diff.]");
      console.log(`  - Local version at: ${oldPath}`);
      console.log(`  + New template at: ${newPath}`);
      console.log(`${"-".repeat(30)}\n`);
      return;
    }

    console.log(`\n--- Diff for ${fileName} ---`);
    let hasDiff = false;

    try {
      execSync(
        process.platform === "win32" ? "where diff" : "command -v diff",
        { stdio: "ignore" },
      );
      hasDiff = true;
    } catch (_e) {
      hasDiff = false;
    }

    if (hasDiff) {
      try {
        const colorFlag = process.stdout.isTTY ? "--color=always" : "";
        execSync(`diff -u ${colorFlag} "${oldPath}" "${newPath}"`, {
          stdio: "inherit",
        });
      } catch (e) {
        if (e.status !== 1) console.log("  (Error running system diff)");
      }
    } else {
      console.log("  [System 'diff' not found. Summary of changes:]");
      console.log(`  - Local version at: ${oldPath}`);
      console.log(`  + New template at: ${newPath}`);
      console.log("  (Install diff or Git to see inline diffs here)");
    }
    console.log(`${"-".repeat(30)}\n`);
  }
}

function parseArgs() {
  const args = process.argv.slice(2);
  const config = { destDir: process.cwd(), force: false, multiLab: false, help: false };

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === "--dest") {
      if (i + 1 >= args.length) throw new Error("Missing value for --dest");
      config.destDir = resolve(args[++i]);
    } else if (arg === "--force" || arg === "-f") {
      config.force = true;
    } else if (arg === "--multi") {
      config.multiLab = true;
    } else if (arg === "--help" || arg === "-h") {
      config.help = true;
    } else {
      throw new Error(`Unknown option: ${arg}`);
    }
  }
  return config;
}

function resolveSourceDir() {
  try {
    return join(
      dirname(require.resolve("@christianmz565/lab-report/package.json")),
      "template",
    );
  } catch (_e) {
    return join(__dirname, "..", "..", "template");
  }
}

async function run() {
  const { destDir, force, multiLab, help } = parseArgs();

  if (help) {
    console.log(HELP_TEXT.trim());
    return;
  }

  const sourceDir = resolveSourceDir();

  if (!existsSync(sourceDir)) {
    throw new Error(`Template source not found at ${sourceDir}`);
  }

  const updater = new Updater(sourceDir, destDir, force, multiLab);
  await updater.processUpdates();
}

run().catch((err) => {
  console.error(err.message ?? err);
  process.exit(1);
});

File: internal/cli/update.go
package cli

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/charmbracelet/huh"
	"github.com/christianmz565/lab-report/internal/diff"
	"github.com/christianmz565/lab-report/internal/templates"
	"github.com/spf13/cobra"
)

type updateOptions struct {
	dest  string
	force bool
	multi bool
	repo  string
	ref   string
}

func newUpdateCmd() *cobra.Command {
	opt := updateOptions{}

	cmd := &cobra.Command{
		Use:   "update",
		Short: "Update an existing template installation",
		RunE: func(cmd *cobra.Command, args []string) error {
			return runUpdate(cmd.Context(), opt)
		},
	}

	cmd.Flags().StringVar(&opt.dest, "dest", "", "Destination directory (default: current working directory)")
	cmd.Flags().BoolVarP(&opt.force, "force", "f", false, "Apply all updates without prompting")
	cmd.Flags().BoolVar(&opt.multi, "multi", false, "Force multi-lab update mode")
	cmd.Flags().StringVar(&opt.repo, "repo", "christianmz565/lab-report", "GitHub repo to fetch templates from (owner/repo)")
	cmd.Flags().StringVar(&opt.ref, "ref", "main", "Git ref to fetch templates from")

	return cmd
}

type updateDecision string

const (
	decYes  updateDecision = "yes"
	decNo   updateDecision = "no"
	decAll  updateDecision = "all"
	decQuit updateDecision = "quit"
)

var imageExt = regexp.MustCompile(`(?i)\.(png|jpe?g|gif|svg|webp|ico)$`)

func runUpdate(ctx context.Context, opt updateOptions) error {
	destDir := opt.dest
	if destDir == "" {
		destDir = "."
	}
	destDir, err := filepath.Abs(destDir)
	if err != nil {
		return err
	}

	cfg, ok, err := ReadConfig(destDir)
	if err != nil {
		fmt.Fprintln(os.Stdout, err.Error())
		// keep going; we'll fall back to flags
	}

	isMulti := opt.multi
	if ok {
		isMulti = isMulti || cfg.MultiLab
	} else {
		// Mimic old behavior: create default config and ask user to re-run.
		defaultCfg := LabReportConfig{MultiLab: isMulti}
		if err := WriteConfig(destDir, defaultCfg); err != nil {
			return err
		}
		return fmt.Errorf("labreport.json not found. Created default config; please verify and run update again")
	}

	owner, repo, err := templates.ParseRepo(opt.repo)
	if err != nil {
		return err
	}
	src := templates.Source{Owner: owner, Repo: repo, Ref: opt.ref}

	ctx, cancel := context.WithTimeout(ctx, 45*time.Second)
	defer cancel()

	remoteFiles, err := templates.Fetch(ctx, src)
	if err != nil {
		// Dev/offline fallback: if running from the repo, use ./template.
		if local, lerr := templates.LoadFromDir("template"); lerr == nil {
			remoteFiles = local
		} else {
			return err
		}
	}
	m, err := templates.LoadManifest(remoteFiles)
	if err != nil {
		// Dev/offline fallback if the remote repo hasn't been updated yet.
		if local, lerr := templates.LoadFromDir("template"); lerr == nil {
			remoteFiles = local
			m, err = templates.LoadManifest(remoteFiles)
		}
		if err != nil {
			return err
		}
	}

	fmt.Fprintf(os.Stdout, "Detected %s setup.\n", map[bool]string{true: "multi-lab", false: "single-lab"}[isMulti])
	fmt.Fprintf(os.Stdout, "Checking for updates in: %s\n\n", destDir)

	if err := maybeRemoveLegacyScripts(destDir, opt.force); err != nil {
		return err
	}

	// .prepare.config is user-owned; ensure it exists but never update it from remote templates.
	prepareCfg := filepath.Join(destDir, ".prepare.config")
	if !FileExists(prepareCfg) {
		if err := WriteFileAtomic(prepareCfg, []byte(""), 0o644); err != nil {
			return err
		}
		fmt.Fprintln(os.Stdout, "Created: .prepare.config")
	}

	entries := buildUpdateEntries(m, isMulti, destDir)
	entries = expandDirEntries(remoteFiles, entries)
	applied := 0
	autoAcceptAll := opt.force

	for _, e := range entries {
		dstPath := filepath.Join(destDir, filepath.FromSlash(e.Dest))
		srcPath := e.Src

		switch e.Kind {
		case templates.KindDir:
			if err := EnsureDir(dstPath); err != nil {
				return err
			}
			continue
		case templates.KindFile:
			remote, ok := remoteFiles[srcPath]
			if !ok {
				continue
			}

			local, lerr := os.ReadFile(dstPath)
			isNew := os.IsNotExist(lerr)
			if lerr != nil && !isNew {
				return lerr
			}
			if !isNew && SameContent(local, remote) {
				continue
			}

			label := e.Dest
			apply := func() error {
				if err := WriteFileAtomic(dstPath, remote, 0o644); err != nil {
					return err
				}
				applied++
				return nil
			}

			if e.AutoUpdate || autoAcceptAll {
				if err := apply(); err != nil {
					return err
				}
				tag := "[FORCED]"
				if e.AutoUpdate {
					tag = "[AUTO-UPDATE]"
				}
				fmt.Fprintf(os.Stdout, "✓ Updated %s: %s\n", tag, label)
				continue
			}

			if isNew {
				fmt.Fprintf(os.Stdout, "\n[NEW FILE] %s\n", label)
			} else {
				showDiff(label, string(local), string(remote), dstPath, srcPath)
			}

			dec, err := promptUpdateDecision(label)
			if err != nil {
				return err
			}
			switch dec {
			case decAll:
				autoAcceptAll = true
				fallthrough
			case decYes:
				if err := apply(); err != nil {
					return err
				}
				fmt.Fprintf(os.Stdout, "✓ Updated: %s\n", label)
			case decNo:
				fmt.Fprintf(os.Stdout, "Skipped: %s\n", label)
			case decQuit:
				fmt.Fprintln(os.Stdout, "\nStopping update process.")
				return nil
			}
		default:
			return fmt.Errorf("unknown manifest entry kind: %s", e.Kind)
		}
	}

	fmt.Fprintf(os.Stdout, "\nUpdate finished. %d files updated.\n", applied)
	return nil
}

func buildUpdateEntries(m *templates.Manifest, isMulti bool, destDir string) []templates.Entry {
	var out []templates.Entry
	add := func(entries ...[]templates.Entry) {
		for _, list := range entries {
			out = append(out, list...)
		}
	}

	if isMulti {
		add(m.Multi.Root)
		out = append(out, m.Multi.Readme)

		labs := detectLabDirs(destDir)
		sort.Strings(labs)
		for _, lab := range labs {
			out = append(out, substituteLab(m.Multi.LabFiles, lab)...)
		}
	} else {
		add(append(m.Common, m.Single...))
	}

	// Dedup by destination path (last write wins).
	seen := map[string]templates.Entry{}
	for _, e := range out {
		seen[e.Dest] = e
	}

	keys := make([]string, 0, len(seen))
	for k := range seen {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	final := make([]templates.Entry, 0, len(keys))
	for _, k := range keys {
		final = append(final, seen[k])
	}
	return final
}

func expandDirEntries(remote templates.Files, entries []templates.Entry) []templates.Entry {
	out := make([]templates.Entry, 0, len(entries))
	for _, e := range entries {
		out = append(out, e)
		if e.Kind != templates.KindDir || strings.TrimSpace(e.Src) == "" {
			continue
		}

		prefix := strings.TrimSuffix(e.Src, "/") + "/"
		paths := make([]string, 0)
		for p := range remote {
			if strings.HasPrefix(p, prefix) {
				paths = append(paths, p)
			}
		}
		sort.Strings(paths)

		for _, p := range paths {
			rel := strings.TrimPrefix(p, prefix)
			if rel == "" {
				continue
			}
			out = append(out, templates.Entry{
				Kind: templates.KindFile,
				Src:  p,
				Dest: filepath.ToSlash(filepath.Join(e.Dest, rel)),
			})
		}
	}
	return out
}

func detectLabDirs(destDir string) []string {
	entries, err := os.ReadDir(destDir)
	if err != nil {
		return nil
	}
	labs := make([]string, 0)
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		name := e.Name()
		if strings.HasPrefix(name, ".") || name == "node_modules" {
			continue
		}
		if FileExists(filepath.Join(destDir, name, "report.typ")) {
			labs = append(labs, name)
		}
	}
	return labs
}

func showDiff(fileName, oldText, newText, oldPath, newSrc string) {
	fmt.Fprintf(os.Stdout, "\n--- Diff for %s ---\n", fileName)
	if imageExt.MatchString(fileName) {
		fmt.Fprintln(os.Stdout, "  [Binary image file changed. Cannot display inline diff.]")
		fmt.Fprintf(os.Stdout, "  - Local version at: %s\n", oldPath)
		fmt.Fprintf(os.Stdout, "  + New template at: %s\n", newSrc)
		fmt.Fprintln(os.Stdout, strings.Repeat("-", 30)+"\n")
		return
	}

	fmt.Fprint(os.Stdout, diff.UnifiedLineDiff(oldText, newText))
	fmt.Fprintln(os.Stdout, strings.Repeat("-", 30)+"\n")
}

func promptUpdateDecision(label string) (updateDecision, error) {
	var choice updateDecision
	form := huh.NewForm(
		huh.NewGroup(
			huh.NewSelect[updateDecision]().
				Title(fmt.Sprintf("Apply update to %s?", label)).
				Options(
					huh.NewOption("Yes", decYes),
					huh.NewOption("No", decNo),
					huh.NewOption("All remaining", decAll),
					huh.NewOption("Quit", decQuit),
				).
				Value(&choice),
		),
	)
	if err := form.Run(); err != nil {
		return "", err
	}
	return choice, nil
}

func maybeRemoveLegacyScripts(destDir string, force bool) error {
	legacy := []string{"prepare.js", "capture.js"}
	found := make([]string, 0)
	for _, f := range legacy {
		p := filepath.Join(destDir, f)
		if FileExists(p) {
			found = append(found, f)
		}
	}
	if len(found) == 0 {
		return nil
	}

	if force {
		for _, f := range found {
			_ = os.Remove(filepath.Join(destDir, f))
			fmt.Fprintf(os.Stdout, "Removed legacy script: %s\n", f)
		}
		return nil
	}

	var confirm bool
	form := huh.NewForm(huh.NewGroup(
		huh.NewConfirm().
			Title("Remove legacy Node scripts (prepare.js/capture.js)?").
			Description("These are no longer needed; use `lab-report prepare` and `lab-report capture` instead.").
			Value(&confirm),
	))
	if err := form.Run(); err != nil {
		return err
	}
	if !confirm {
		return nil
	}

	for _, f := range found {
		if err := os.Remove(filepath.Join(destDir, f)); err != nil {
			return err
		}
		fmt.Fprintf(os.Stdout, "Removed legacy script: %s\n", f)
	}
	return nil
}
